### PR TITLE
Make tuple types work in autodiff.

### LINKED ISSUE #4891: Implement autodiff support for Tuple types.

### PR DESCRIPTION
**Pull request by csyonghe**
Closes #4891.

- Add `extension` declaration in stdlib for tuple types.
- Implement differentiation logic for tuple types, both in the frontend and in the IR.
- Make sure UpdateElement access keys on tuple types are lowered to struct keys when lowering tuple types.

### CODE REVIEW

File: tests/language-feature/tuple/tuple-autodiff.slang [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+[numthreads(1, 1, 1)]
+void computeMain(uint3 dispatchThreadID : SV_DispatchThreadID)
+{
+    A a = makeTuple(1.0, makeTuple(1.0, 2.0));
+
+    var dpa = diffPair(a);
+
+    A.Differential dout = makeTuple(1.0, makeTuple(1.0, 1.0));
+    
+    __bwd_diff(f)(dpa, dout);
```
Comment (saipraveenb25) [inline tests/language-feature/tuple/tuple-autodiff.slang:44]: switch to `bwd_diff`?

**Review by saipraveenb25 — DISMISSED** [review]
Everything looks good to me!

We do eventually need to figure out how to do optional constraints...

File: source/slang/slang-ir-autodiff.cpp [hunk 2]
```diff
@@ -448,26 +448,27 @@ void DifferentiableTypeConformanceContext::setFunc(IRGlobalValueWithCode* func)
...
-
-                    for (UInt i = 0; i < concreteType->getOperandCount(); i++)
+                    if (auto witnessPack = as<IRMakeWitnessPack>(witness))
                     {
-                        auto element = concreteType->getOperand(i);
-                        auto elementWitness = witnessPack->getOperand(i);
-                        differentiableWitnessDictionary.addIfNotExists(
-                            (IRType*)element,
-                            _lookupWitness(&subBuilder, elementWitness, sharedContext->differentialAssocTypeStructKey));
+                        SLANG_ASSERT(witnessPack);
```
Comment (jkwak-work) [inline source/slang/slang-ir-autodiff.cpp:456]: This assert is no longer needed.

**Review by jkwak-work — DISMISSED** [review]
Looks good to me.

**Review by csyonghe — COMMENTED** [review]
Address reviews.

**Pull request merged**